### PR TITLE
Bump Scikit Learn Interface & Speed Improvement Using pypuf

### PR DESCRIPTION
**Pull request by nils-wisiol**
This PR has two parts:

1. The first part uses the current directory instead of a hard coded path and adjusts the interface to scikit-learn. Both are required to run this code on a new installation.
2. The second part updates the CRP generation in the repository to use pypuf instead of the delay-based simulation that is in this codebase. I included pytest test routines to show that the simulation result is identical.
I conditioned the test for duplicates in the random challenges on the number of stages, as for more than 32 stages the chance that we have duplicate challenges in any digestible-size challenge set is practically zero. 
The simulation speed is improved by several orders of magnitude, on my `Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz` it now takes less than 5 seconds to generate a million CRPs of an 64-bit 8-XOR Arbiter PUF. This could be further improved by modifying the simulation for XORArbiterPUFs to not simulate the chains separately.

Unfortunately, using the code from this PR, both including only the first part and including both parts, I cannot reproduce the results for 64-bit 4-XOR Arbiter PUFs from the paper running `python main.py --streams 4 --stages 64 --challenges 200000`. After about 5mins, it stops with accuracy 50%. I'm fairly confident that the simulation improvement isn't to blame, as the tests show that the simulation results didn't change. Maybe scikit-learn behavior changed since publication? I'm using scikit-learn 0.24.1 from pypi. Running older versions failed due to different reasons such as installation failure or incompatible API.

@kmursi

### CODE REVIEW

**Comment by nils-wisiol** [comment]
d17d41a was able to train to 98% accuracy using 2M CRPs (`python main.py --streams 4 --stages 64 --challenges 2000000`) and scikit-learn 0.24.1.
